src/main.lua

```diff
@@ -1335,41 +1343,59 @@ addEventingHandler("totalTokenSupply", utils.hasMatchingTag("Action", "Total-Tok
 	local delegatedSupply = 0
 	local withdrawSupply = 0
 	local protocolBalance = balances.getBalance(Protocol)
-	local balances = balances.getBalances()
-	for _, balance in pairs(balances) do
-		totalSupply = totalSupply + balance
+	local userBalances = balances.getBalances()
+
+	-- tally circulating supply
+	for _, balance in pairs(userBalances) do
 		circulatingSupply = circulatingSupply + balance
 	end
-	-- gateways and delegates
+	totalSupply = totalSupply + circulatingSupply
+
+	-- tally supply stashed in gateways and delegates
 	local gateways = gar.getGateways()
 	for _, gateway in pairs(gateways) do
 		totalSupply = totalSupply + gateway.operatorStake + gateway.totalDelegatedStake
 		stakedSupply = stakedSupply + gateway.operatorStake
 		delegatedSupply = delegatedSupply + gateway.totalDelegatedStake
 		for _, delegate in pairs(gateway.delegates) do
-			-- check vaults
+			-- tally delegates' vaults
 			for _, vault in pairs(delegate.vaults) do
 				totalSupply = totalSupply + vault.balance
 				withdrawSupply = withdrawSupply + vault.balance
 			end
 		end
-		-- iterate through vaults
+		-- tally gateway's own vaults
 		for _, vault in pairs(gateway.vaults) do
 			totalSupply = totalSupply + vault.balance
 			withdrawSupply = withdrawSupply + vault.balance
 		end
 	end
 
-	-- vaults
-	local vaults = vaults.getVaults()
-	for _, vaultsForAddress in pairs(vaults) do
+	-- user vaults
+	local userVaults = vaults.getVaults()
+	for _, vaultsForAddress in pairs(userVaults) do
 		-- they may have several vaults iterate through them
 		for _, vault in pairs(vaultsForAddress) do
 			totalSupply = totalSupply + vault.balance
 			lockedSupply = lockedSupply + vault.balance
 		end
 	end
 
+	lastKnownTotalSupply = totalSupply
+	lastKnownCirculatingSupply = circulatingSupply
+	lastKnownLockedSupply = lockedSupply
+	lastKnownStakedSupply = stakedSupply
+	lastKnownDelegatedSupply = delegatedSupply
+	lastKnownWithdrawSupply = withdrawSupply
+
+	msg.ioEvent:addField("Total-Token-Supply", lastKnownTotalSupply)
+	msg.ioEvent:addField("Circulating-Supply", lastKnownCirculatingSupply)
+	msg.ioEvent:addField("Locked-Supply", lastKnownLockedSupply)
+	msg.ioEvent:addField("Staked-Supply", lastKnownStakedSupply)
+	msg.ioEvent:addField("Delegated-Supply", lastKnownDelegatedSupply)
+	msg.ioEvent:addField("Withdraw-Supply", lastKnownWithdrawSupply)
+	msg.ioEvent:addField("Protocol-Balance", protocolBalance)
```

Comment:
FWIW - this is mostly used as a dryRun only handler - so it's unlikely to ever surface event data. that being said, we can move the computation to a separate function and call it before/after every write message and add the event data you've added here

Reply:
I was toying with factoring this out and calling it any time functions that affect the supplies get called but I worry that it could get unnecessarily computationally expensive at scale (also from a gas perspective). I haven't ruled it out for ticks so that we get it all at least once a day.